addons/onboarding/models/onboarding_onboarding.py

```diff
@@ -41,15 +46,20 @@ class Onboarding(models.Model):
         ('route_name_uniq', 'UNIQUE (route_name)', 'Onboarding alias must be unique.'),
     ]
 
+    @api.constrains('is_per_company', 'step_ids')
+    def check_is_per_company_consistency(self):
+        for onboarding in self:
+            if any(is_step_per_company != onboarding.is_per_company
```

Comment:
Or
```python
if set(onboarding.step_ids.mapped('is_per_company')) != {onboarding.is_per_company}:
```

If you worry about the performances (you want to fail early and not iterate over the whole recordset), don't use mapped or filtered at all. You iterate over all the steps when there is no error though, so I think it is overkill.
```python
if any(
    step.is_per_company != onboarding.is_per_company
    for step in onboarding.step_ids
):
```

Reply:
> This could be slower as we'll be comparing **all** records before returning anything. This thinking makes me want to use something like the longer (for efficiency and user feedback, as we already do in other modules):
> 
> ```python
> invalid_onboardings = self.browse()
> for onboarding in self:
>     for step in onboarding.step_ids:
>         if step.is_per_company != onboarding.is_per_company:
>             invalid_onboardings |= onboarding
>             continue
> if invalid_onboardings:
>     raise ValidationError(_('Onboarding and step `is_per_company` must be equal, '
>                             'which is not the case in the following onboardings:')
>                           + '\n- ' + '\n- '.join(invalid_onboardings.mapped('name')))
> ```

Your `continue` is useless, it is the last statement of the containing `for` loop.
Also, since you are fetching all the `invalid_onboardings`, it doesn't improve the time complexity, but the error message.

Your previous implementation was also in the same complexity because of the `mapped`; and like I said, since you will eventually need to check all the records when there are no errors, we don't really care about failing early in case of an error: this is a pretty inexpensive check to do compared to the rest.